pkg/controller/nodeipam/ipam/multicidrset/cidr_set.go

```diff
@@ -0,0 +1,363 @@
+limitations under the License.
+*/
+
+package multicidrset
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+	"math/bits"
+	"net"
+	"sync"
+)
+
+// CidrSet manages a set of CIDR ranges from which blocks of IPs can
+// be allocated from.
+type CidrSet struct {
+	sync.Mutex
+	// clusterCIDR is the CIDR assigned to the cluster.
+	clusterCIDR *net.IPNet
+	// clusterMaskSize is the mask size, in bits, assigned to the cluster.
+	// caches the mask size to avoid the penalty of calling clusterCIDR.Mask.Size().
+	clusterMaskSize int
+	// nodeMask is the network mask assigned to the nodes.
+	nodeMask net.IPMask
+	// nodeMaskSize is the mask size, in bits,assigned to the nodes
+	// caches the mask size to avoid the penalty of calling nodeMask.Size().
+	nodeMaskSize int
+	// maxCIDRs is the maximum number of CIDRs that can be allocated.
+	maxCIDRs int
+	// allocatedCIDRs counts the number of CIDRs allocated.
+	allocatedCIDRs int
+	// nextCandidate points to the next CIDR that should be free.
+	nextCandidate int
+	// label is used to identify the metrics.
+	label string
+	// terminating is used to identify whether cidrSet has been marked for termination.
+	terminating bool
+	// AllocatedCIDRMap stores all the allocated CIDRs from the current CIDRSet.
+	AllocatedCIDRMap map[string]bool
+}
+
+// ClusterCIDR is an internal representation of the ClusterCIDRConfig object.
+type ClusterCIDR struct {
+	Name        string
+	IPv4CIDRSet *CidrSet
+	IPv6CIDRSet *CidrSet
+	// AssociatedNodes is used to identify which nodes have CIDRs allocated from this ClusterCIDR.
+	AssociatedNodes map[string]bool
+	// Terminating is used to identify whether ClusterCIDR has been marked for termination.
+	Terminating bool
+}
+
+const (
+	// The subnet mask size cannot be greater than 16 more than the cluster mask size
+	// TODO: https://github.com/kubernetes/kubernetes/issues/44918
+	// clusterSubnetMaxDiff limited to 16 due to the uncompressed bitmap.
+	// Due to this limitation the subnet mask for IPv6 cluster cidr needs to be >= 48
+	// as default mask size for IPv6 is 64.
+	clusterSubnetMaxDiff = 16
+	// halfIPv6Len is the half of the IPv6 length.
+	halfIPv6Len = net.IPv6len / 2
+)
+
+var (
+	// ErrCIDRRangeNoCIDRsRemaining occurs when there is no more space
+	// to allocate CIDR ranges.
+	ErrCIDRRangeNoCIDRsRemaining = errors.New(
+		"CIDR allocation failed; there are no remaining CIDRs left to allocate in the accepted range")
+	// ErrCIDRSetSubNetTooBig occurs when the subnet mask size is too
+	// big compared to the CIDR mask size.
+	ErrCIDRSetSubNetTooBig = errors.New(
+		"New CIDR set failed; the node CIDR size is too big")
+)
+
+// NewCIDRSet creates a new CidrSet.
+func NewCIDRSet(clusterCIDR *net.IPNet, subNetMaskSize int) (*CidrSet, error) {
+	clusterMask := clusterCIDR.Mask
+	clusterMaskSize, bits := clusterMask.Size()
+
+	var maxCIDRs int
+	if (clusterCIDR.IP.To4() == nil) && (subNetMaskSize-clusterMaskSize > clusterSubnetMaxDiff) {
```

Comment:
Is there never a size check for v6?

Reply:
Actually this size check is only for v6, `clusterCIDR.IP.To4() == nil` when it's not v4

Replicated the code from: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodeipam/ipam/cidrset/cidr_set.go#L78

I think the check was initially added due to: https://github.com/kubernetes/kubernetes/issues/44918

Now that we have moved away from bitmap for storing the allocated CIDRs, I think this check is not required anymore? cc: @aojea